refactor: `FeedbackDataset.records` are now `FeedbackRecord` instead of Python `dict`

alvarobartt opened the pull request:
# Description

As of a recent discussion with @frascuchon we should return the `FeedbackRecord` `pydantic.BaseModel` itself, instead of the Python dict conversion of it. So this PR basically removes all the `.dict()` conversion towards the Python client, while it's kept in the API. Besides that we also use `pydantic.BaseModel` validation instead of using `.construct` when there are nested `pydantic.BaseModels`.

Docs have been updated accordingly at #3036 

**Type of change**

- [X] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] Refactor (change restructuring the codebase without changing functionality)
- [X] Documentation update

**How Has This Been Tested**

- [X] Updated unit tests accordingly

**Checklist**

- [X] I have merged the original branch into my forked branch
- [X] I added relevant documentation
- [X] follows the style guidelines of this project
- [X] I did a self-review of my code
- [X] I made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works

alvarobartt committed:
refactor: `records` contains `FeedbackRecord`s

alvarobartt committed:
fix: skip `construct` when there are nested `pydantic.BaseModel`s

alvarobartt committed:
fix: unit tests when accessing `records`

alvarobartt committed:
docs: update docstrings

alvarobartt committed:
fix: remove `.dict()` in `from_huggingface`

alvarobartt committed:
fix: missing `dict()` converstion in `test_get_records`

frascuchon commented on src/argilla/client/feedback.py:
```diff
@@ -91,16 +90,7 @@ def responses_must_be_a_list(cls, v):
-    external_id: Optional[str] = None
-    inserted_at: Optional[str] = None
-    updated_at: Optional[str] = None
+        extra = Extra.ignore
```
Can we show a warning if more properties than expected are passed?  Just to let it know  to the user

alvarobartt replied:
I don't think we can combine `Extra.ignore` with a `root_validator` to make sure that just the existing fields are provided, maybe @gabrielmbmb knows a cleaner way to do so?

alvarobartt replied:
Also note that we want to ignore the extras, also the ones that come from Argilla for the moment i.e. `updated_at`, `created_at`, etc. fields that are not required/needed by the user, also to keep consistency for the records. So when we log records to Argilla and then we want to pull them, just the `FeedbackRecord` fields are kept from the API response

frascuchon commented on src/argilla/client/feedback.py:
```diff
@@ -412,7 +402,7 @@ def fetch_records(self) -> None:
                 client=httpx_client, id=self.argilla_id, offset=0, limit=FETCHING_BATCH_SIZE
             ).parsed
-            self.__records = first_batch.items
+            self.__records = [FeedbackRecord(**item.dict()) for item in first_batch.items]
```
Should we use `FeedbackRecord.parse_obj` instead? Not sure what's the difference

alvarobartt replied:
So we can either `FeedbackRecord(**item.dict())` or `FeedbackRecord.parse_obj(item.dict())`, it's the same thing in this case

frascuchon replied:
I realize that `item` is an object. Should we use [`from_orm`](https://docs.pydantic.dev/latest/usage/models/#orm-mode-aka-arbitrary-class-instances)?  So, we don't need to convert them to dict.

alvarobartt replied:
Sure we can also do that, but we need to update the `Config` then, right?

alvarobartt replied:
But I like this to avoid converting to `dict()` in between whenever we receive `pydantic.BaseModel`s already

alvarobartt replied:
I've used [`parse_obj_as`](https://docs.pydantic.dev/latest/usage/models/#parsing-data-into-a-specified-type), as we're moving from a `pydantic.BaseModel` to another `pydantic.BaseModel`, not from a generic class, so I think this may be more aligned to what we want 😄

frascuchon commented on src/argilla/client/feedback.py:
```diff
@@ -426,7 +416,7 @@ def fetch_records(self) -> None:
                         limit=FETCHING_BATCH_SIZE,
                     ).parsed
-                    records = batch.items
+                    records = [FeedbackRecord(**item.dict()) for item in batch.items]
```
ditto

alvarobartt replied:
Same here, `parse_obj_as` has been used instead!

frascuchon commented on src/argilla/client/feedback.py:
```diff
@@ -484,7 +474,7 @@ def add_records(
             ... )
             >>> dataset.records
-            [FeedbackRecord(fields={"text": "This is the first record", "label": "positive"}, responses=[{"user_id": None, "values": {"question-1": {"value": "This is the first answer"}, "question-2": {"value": 5}}}])]
+            [FeedbackRecord(fields={"text": "This is the first record", "label": "positive"}, responses=[ResponseSchema(user_id=None, values={"question-1": ValueSchema(value="This is the first answer"), "question-2": ValueSchema(value=5)})])]
```
same change here, since the `user_id` is an optional value and we're setting it to `None` in the example, makes sense to remove it from the response initialization.

alvarobartt replied:
So the thing is that since it's optional it's set to `None` by default, so when we print the `FeedbackRecord` we see the `responses` as a `ResponseSchema` with all its values, so I think it's nice that we keep it as-is in the docstrings, as that's what you get when you `print(self.records)`, and also to give the users visibility about the `user_id` field in `ResponseSchema`

frascuchon commented on src/argilla/client/feedback.py:
```diff
@@ -498,20 +488,18 @@ def add_records(
         for record in records:
             try:
-                record.fields = self.__fields_schema.parse_obj(record.fields)
+                self.__fields_schema.parse_obj(record.fields)
```
This line, is just to apply the fields_schema validation, right? I understand that fields will be accessed as dicts instead of class instances:

```python
records.fields["text"]
```
instead of 
```python
records.fields.text
```

alvarobartt replied:
Yes, since before we were mapping the `FeedbackRecord.fields` to `FieldsSchema`, but we were calling `construct` over `FeedbackRecord` instead of creating it as `FeedbackRecord(...)`, so it was still a Python dict, adding that doesn't make too much sense for the moment, so we're just validating the `fields` based on `FeedbackDataset.fields` auto-generated `pydantic` schema

davidberenstein1957 commented on src/argilla/client/feedback.py:
```diff
@@ -91,16 +90,7 @@ def responses_must_be_a_list(cls, v):
-        extra = Extra.forbid
-
-
-class OfflineFeedbackRecord(BaseModel):
```
Do all Pydantic models doo validation during attribute assignment?

alvarobartt replied:
Yes they do, unless you call `.construct()`, that just builds it as a `pydantic.BaseModel` ignoring the nested `pydantic.BaseModel`s and assuming those are Python dicts, which is not what we want here. So instantiating or calling `parse_obj` over a `pydantic.BaseModel` does validate it 😄

davidberenstein1957 commented on src/argilla/client/feedback.py:
```diff
@@ -484,7 +474,7 @@ def add_records(
-            [FeedbackRecord(fields={"text": "This is the first record", "label": "positive"}, responses=[{"user_id": None, "values": {"question-1": {"value": "This is the first answer"}, "question-2": {"value": 5}}}])]
+            [FeedbackRecord(fields={"text": "This is the first record", "label": "positive"}, responses=[ResponseSchema(user_id=None, values={"question-1": ValueSchema(value="This is the first answer"), "question-2": ValueSchema(value=5)})])]
         """
         if isinstance(records, list):
```
I think `elif` statements would be better suited for the other cases

alvarobartt replied:
Those are non-colliding if statements, but sure, I can make those `elif` instead, even though there's not a clear `else` condition, should we raise `ValueError`?

frascuchon commented on src/argilla/client/sdk/v1/datasets/api.py:
```diff
@@ -207,7 +207,7 @@ def get_records(
 
     if response.status_code == 200:
-        parsed_response = FeedbackRecordsModel.construct(**response.json())
+        parsed_response = FeedbackRecordsModel(**response.json())
```
`parse_obj` like?

alvarobartt replied:
Same thing as above, we removed the usage of `.construct` as that's not doing the validation and the nested `pydantic.BaseModel`s are lost and transformed into Python dictionaries, but here we could either do the current or `FeedbackRecordsModel.parse_obj(response.json())`, the outcome will be the same

frascuchon commented on src/argilla/client/sdk/v1/datasets/api.py:
```diff
@@ -286,7 +286,7 @@ def get_fields(
 
     if response.status_code == 200:
-        parsed_response = [FeedbackFieldModel.construct(**item) for item in response.json()["items"]]
+        parsed_response = [FeedbackFieldModel(**item) for item in response.json()["items"]]
```
ditto

alvarobartt committed:
fix: raise `ValueError` in `add_records`

Raised if the provided records are neither a list of `FeedbackRecord` or dict, a `FeedbackRecord`, or a dict

Co-authored-by: david <david.m.berenstein@gmail.com>

alvarobartt committed:
fix: `add_records` when `list` provided

alvarobartt committed:
fix: raise `ValueError` if `records=[]`

alvarobartt committed:
test: add `add_records` integration tests

alvarobartt committed:
perf: use `parse_obj_as` to avoid `dict` conversion
